docs/jax-101/02-jitting.md

```diff
@@ -0,0 +1,293 @@
+
+When tracing, JAX wraps each argument by a *tracer* object. These tracers then record all JAX operations performed on them during the function call (which happens in regular Python). Then, JAX uses the tracer records to reconstruct the entire function. The output of that reconstruction is the jaxpr. Since the tracers do not record the Python side-effects, they do not appear in the jaxpr. However, the side-effects still happen during the trace itself.
+
+Note: the Python `print()` function is not pure: the text output is a side-effect of the function. Therefore, any `print()` calls will only happen during tracing, and will not appear in the jaxpr:
+
+```{code-cell} ipython3
+:id: JxV2p7e2RawC
+:outputId: af7472de-df57-412f-83cb-e468c4b02d64
+
+def log2_with_print(x):
+  print("printed x:", x)
+  ln_x = jnp.log(x)
+  ln_2 = jnp.log(2.0)
+  return ln_x / ln_2
+
+print(jax.make_jaxpr(log2_with_print)(3.))
+```
+
++++ {"id": "f6W_YYwRRwGp"}
+
+See how the printed `x` is a `Traced` object? That's the JAX internals at work.
+
+The fact that the Python code runs at least once is strictly an implementation detail, and so shouldn't be relied upon. However, it's useful to understand as you can use it when debugging to print out intermediate values of a computation.
+
++++ {"id": "PgVqi6NlRdWZ"}
+
+A key thing to understand is that jaxpr captures the function as executed on the parameters given to it. For example, if we have a conditional, jaxpr will only know about the branch we take:
+
+```{code-cell} ipython3
+:id: hn0CuphEZKZm
+:outputId: 669c5880-790d-4e5c-b635-f3acc21a0e02
+
+def log2_if_rank_2(x):
+  if x.ndim == 2:
+    ln_x = jnp.log(x)
+    ln_2 = jnp.log(2.0)
+    return ln_x / ln_2
+  else:
+    return x
+
+print(jax.make_jaxpr(log2_if_rank_2)(jax.numpy.array([1, 2, 3])))
+```
+
++++ {"id": "Qp3WhqaqvHyD"}
+
+## JIT compiling a function
+
+As explained before, JAX enables operations to execute on CPU/GPU/TPU using the same code.
+Let's look at an example of computing a *Scaled Exponential Linear Unit*
+([SELU](https://proceedings.neurips.cc/paper/6698-self-normalizing-neural-networks.pdf)), an
+operation commonly used in deep learning:
+
+
+```{code-cell} ipython3
+:id: JAXFYtlRvD6p
+:outputId: e850862d-c9b4-4cfe-85a2-fb707178a957
+
+import jax
+import jax.numpy as jnp
+
+def selu(x, alpha=1.67, lambda_=1.05):
+  return lambda_ * jnp.where(x > 0, x, alpha * jnp.exp(x) - alpha)
+
+x = jnp.arange(1000000)
+%timeit selu(x).block_until_ready()
+```
+
++++ {"id": "ecN5lEXe6ncy"}
+
+The code above is sending one operation at a time to the accelerator. This limits the ability of the XLA compiler to optimize our functions.
+
+Naturally, what we want to do is give the XLA compiler as much code as possible, so it can fully optimize it. For this purpose, JAX provides the `jax.jit` transformation, which will Just In Time (JIT) compile a JAX-compatible function. The example below shows how to use JIT to speed up the previous function.
```

Comment:
Suggestion:

- Similar to _"Every JAX operation is eventually expressed in terms of these fundamental XLA operations, which is what enables just-in-time (JIT) compilation."_ in https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html

```suggestion
Naturally, what we want to do is give the XLA compiler as much code as possible, so it can fully optimize it. For this purpose, JAX provides the `jax.jit` transformation, which will Just-In-Time (JIT) compile a JAX-compatible function. The example below shows how to use JIT to speed up the previous function.
```

🤷  cc @skye